### ISSUE TITLE
building `romcal`

### ISSUE THREAD

**Issue by tukusejssirs** [romcal/romcal#49]
At first, I wanted to build something like in [`romcal v1.2.0`](https://github.com/pejulian/romcal/releases) release—there’s an option to build and run it in BASH. I wanted to test the current (unreleased) version of the `romcal` code from `romcal-es6` branch, but I couldn’t find out how.

Then, I’ve found out that [it could be built using `npm`](https://github.com/pejulian/romcal/tree/romcal-es6#builds) and with a web browser (e.g. `firefox`), so I tried it. Here is what I tried on Kubuntu 18.04 AMD64:
```bash
    $ sudo apt-get install npm
    # Output removed
    $ sudo npm i -g npm
    # Output removed
    $ git clone git@github.com:pejulian/romcal.git -b romcal-es6 romcal_bt2
    Cloning into 'romcal_bt2'...
    remote: Counting objects: 2125, done.
    remote: Total 2125 (delta 0), reused 0 (delta 0), pack-reused 2125
    Receiving objects: 100% (2125/2125), 2.21 MiB | 836.00 KiB/s, done.
    Resolving deltas: 100% (1552/1552), done.
    $ cd romcal_bt2/
    $ sudo npm install webpack webpack-cli
    npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.1.3 (node_modules/fsevents):
    npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
    
    + webpack-cli@3.0.7
    + webpack@4.12.0
    added 685 packages from 668 contributors and audited 5966 packages in 19.346s
    found 13 vulnerabilities (2 low, 9 moderate, 2 high)
      run `npm audit fix` to fix them, or `npm audit` for details
    ts@ts-kubuntu:~/romcal_bt2$ sudo npm audit fix
    npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
    npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
    
    up to date in 6.011s
    fixed 13 of 13 vulnerabilities in 5966 scanned packages
    $ sudo npm run build
    
    > romcal@1.3.0 build /home/ts/romcal_bt2
    > webpack --config webpack.config.js
    
    Hash: 5c3dad39147a9713ee0c
    Version: webpack 4.12.0
    Time: 587ms
    Built at: 15/06/2018 19:41:27                                                                                                                                                                   
                   Asset     Size  Chunks             Chunk Names                                                                                                                                   
    romcal.bundle.min.js  8.4 KiB  romcal  [emitted]  romcal
    [./src/index.mjs] 3.81 KiB {romcal} [built]
    [0] multi ./src/index.mjs 28 bytes {romcal} [built]
    $ sudo npm install --save-dev webpack
    npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
    npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
    
    + webpack@4.12.0
    updated 1 package and audited 5626 packages in 7.085s
    found 0 vulnerabilities
    
    $ sudo npm build romcal
    npm ERR! path /home/ts/romcal_bt2/romcal/package.json
    npm ERR! code ENOENT
    npm ERR! errno -2
    npm ERR! syscall open
    npm ERR! enoent ENOENT: no such file or directory, open '/home/ts/romcal_bt2/romcal/package.json'
    npm ERR! enoent This is related to npm not being able to find a file.
    npm ERR! enoent 
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/ts/.npm/_logs/2018-06-15T18_07_32_872Z-debug.log
    
    $ cat /home/ts/.npm/_logs/2018-06-15T18_07_32_872Z-debug.log 
    0 info it worked if it ends with ok
    1 verbose cli [ '/usr/bin/node', '/usr/local/bin/npm', 'build', 'romcal' ]
    2 info using npm@6.1.0
    3 info using node@v8.10.0
    4 info build /home/ts/romcal_bt2/romcal
    5 verbose stack Error: ENOENT: no such file or directory, open '/home/ts/romcal_bt2/romcal/package.json'
    6 verbose cwd /home/ts/romcal_bt2
    7 verbose Linux 4.15.0-22-generic
    8 verbose argv "/usr/bin/node" "/usr/local/bin/npm" "build" "romcal"
    9 verbose node v8.10.0
    10 verbose npm  v6.1.0
    11 error path /home/ts/romcal_bt2/romcal/package.json
    12 error code ENOENT
    13 error errno -2
    14 error syscall open
    15 error enoent ENOENT: no such file or directory, open '/home/ts/romcal_bt2/romcal/package.json'
    16 error enoent This is related to npm not being able to find a file.
    17 verbose exit [ -2, true ]
```

This did not look good, however I couldn’t solve the errors. Nevertheless, I tried and opened the `index.html` file in `firefox`and opened the web console. There were the following errors:
```bash
ReferenceError: exports is not defined [Learn More]    romcal.bundle.min.js%20line%2098%20%3E%20eval:4:1
<anonymous>    webpack:///./src/index.mjs?:4:1
./src/index.mjs    file:///home/ts/romcal_bt2/dist/romcal.bundle.min.js:98:1
__webpack_require__    file:///home/ts/romcal_bt2/dist/romcal.bundle.min.js:20:12
<anonymous>    webpack:///multi_./src/index.mjs?:1:18
[0]    file:///home/ts/romcal_bt2/dist/romcal.bundle.min.js:110:1
__webpack_require__    file:///home/ts/romcal_bt2/dist/romcal.bundle.min.js:20:12
<anonymous>     file:///home/ts/romcal_bt2/dist/romcal.bundle.min.js:84:18
<anonymous>     file:///home/ts/romcal_bt2/dist/romcal.bundle.min.js:1:11

ReferenceError: Romcal is not defined [Learn More]    index2.html:28:11
<anonymous>    file:///home/ts/romcal_bt2/index2.html:28:11
```

Btw, I think it would be better to use `document.write()` instead of `console.log()` (or both)—that way the output would be displayed in the `<body>` of the website. It might be even processed and styled with CSS.

So the question is: how could I test the `romcal`? How could I build it from source and either run it in BASH or in a web browser?

**Comment by tukusejssirs** [comment]
I probably found the source of this issue, although I still need to verify it.

**Comment by tukusejssirs** [comment]
## Problem definition

The problem lays in location of `package.json`. The file is in `[romcal-es6 root]/`, however `npm build romcal` checks it in `[romcal-es6 root]/romcal/`.

## Workaround

So, here’s the solution with all the steps required to build `romcal` from source.

```bash
# Note that on Termux, all commands should be run without `sudo`

# Clone `romcal-es6`
git clone git@github.com:pejulian/romcal.git -b romcal-es6 romcal-es6

# Change the working directory to `romcal.es6` root
cd romcal-es6

# Create `romcal` directory
mkdir romcal

# Create symlink to `package.json` in `romcal` directory
ln -s $PWD/package.json $PWD/romcal/package.json

# Install `npm` if it is not yet installed
sudo apt-get -yq update
sudo apt-get -yq install npm

# Install `romcal` dependencies (all dependencies are found at the end of the `package.json` file except for `webpack-cli` and `acorn-dynamic-import`
sudo npm install --save-dev @babel/cli @babel/core @babel/plugin-proposal-class-properties @babel/plugin-proposal-object-rest-spread @babel/preset-env @babel/register @std/esm babel-loader eslint eslint-plugin-node mocha should source-map-loader webpack webpack-cli acorn-dynamic-import

# Build npm
sudo npm run build

# Build romcal
sudo npm build romcal
```

## Solution

Change the path where the `npm` looks for `package.json`. I have no idea how to accomplish it (yet).

---------

PS—I tested it on Linux (Ubuntu 18.04 AMD64) and Android (in Termux ARM64).

PS 2—I suggest to add this to the documentation. For now it would be nice to have documentation in the wiki.

**Comment by tukusejssirs** [comment]
My currect `romcal` build Guide for Linux (using `apt-get` to install `npm`):
```bash
# Install `npm`
sudo apt-get install npm
# Update `npm`
sudo npm i -g npm
# Clone `romcal-es6` (use eith ssh or https to clone)
git clone git@github.com:romcal/romcal.git -b romcal-es6 romcal-es6
#git clone https://github.com/romcal/romcal.git -b romcal-es6 romcal-es6

# Enter the repo root directory
cd romcal-es6
# Install the dependencies locally
# Note: use *must* be in the repo root directory
sudo npm install --save-dev acorn @babel/cli @babel/core \
@babel/plugin-proposal-class-properties @babel/plugin-proposal-object-rest-spread \
@babel/preset-env @babel/register @std/esm babel-loader eslint eslint-plugin-node \
mocha should source-map-loader webpack webpack-cli
# Fix some vulnerabilities by updating the necessary packages
sudo npm audit fix
# Create `[repo_root]/romcal` to fix a partial issue
mkdir -p romcal
# Copy `[repo_root]/package.json` to `[repo_root]/romcal/` to fix the partial issue
cp package.json romcal/
# Actual build of `romcal`
sudo npm build romcal
sudo npm run build
```

However, it still doesn’t work; here are the Web Console errors after opening the `[repo_root]/index.html` in Firefox:

```
ReferenceError: exports is not defined[Learn More]    romcal.bundle.min.js line 98 > eval:4:1
  <anonymous>          webpack:///./src/index.mjs?:4
  mjs                  file:///home/ts/dtp/romcal-es6/dist/romcal.bundle.min.js:98
  __webpack_require__  file:///home/ts/dtp/romcal-es6/dist/romcal.bundle.min.js:20
  <anonymous>          webpack:///multi_./src/index.mjs?:1
  0                    file:///home/ts/dtp/romcal-es6/dist/romcal.bundle.min.js:110
  __webpack_require__  file:///home/ts/dtp/romcal-es6/dist/romcal.bundle.min.js:20
  <anonymous>          file:///home/ts/dtp/romcal-es6/dist/romcal.bundle.min.js:84
  <anonymous>          file:///home/ts/dtp/romcal-es6/dist/romcal.bundle.min.js:1 

ReferenceError: Romcal is not defined[Learn More] index.html:33:11
  <anonymous>        file:///home/ts/dtp/romcal-es6/index.html:33 

The Components object is deprecated. It will soon be removed. index.html
```

**Comment by emagnier** [comment]
This issue is related to the `romcal-es6` branch and is now fixed (#82).
